### ISSUE TITLE
[Suggestion] Improve 'The default constructor is already defined'

### ISSUE THREAD

**Issue by bernaferrari** [dart-lang/sdk#46853]
I tried making a secondary constructor and got confused with the error message.

```dart
class Pixel {
  final int x;
  final int y;

  const Pixel({required this.x, required this.y});

  const from(this.x, this.y); // <-- ERROR, DEFAULT IS ALREADY DEFINED
}
```

Ideally, there could be a quick fix or a suggestion like "try doing Pixel.from" instead.
 I know the default is already defined, but I have no idea how to define a secondary  constructor in dart.

**Comment by bwilkerson** [comment]
The master branch no longer produces the diagnostic you're currently seeing, but I'm not sure the current state is better. The parser is currently assuming that this is an instance method rather than a constructor. As a result we're currently getting the following diagnostics reported:
```
error: Initializing formal parameters can only be used in constructors. (field_initializer_outside_constructor at [test_package] lib/test.dart:7)
error: 'SecondaryConstructor' must have a method body because 'Pixel' isn't abstract. (concrete_class_with_abstract_member at [test_package] lib/test.dart:7)
error: Getters, setters and methods can't be declared to be 'const'. (const_method at [test_package] lib/test.dart:7)
error: Field formal parameters can only be used in a constructor. (field_initializer_outside_constructor at [test_package] lib/test.dart:7)
error: Initializing formal parameters can only be used in constructors. (field_initializer_outside_constructor at [test_package] lib/test.dart:7)
error: Field formal parameters can only be used in a constructor. (field_initializer_outside_constructor at [test_package] lib/test.dart:7)
```

If nothing else, this is very poor recovery on the parser's part. I don't know how hard it would be to change the parser, but it would be nice if it could decide that this is a constructor (based on the presence of some or all of the `const` keyword, the initializing formals, and the empty body) and report that the name is invalid.

**Comment by bernaferrari** [comment]
I like in Rust that sometimes it gives suggestions like "are you trying to do this?". I stopped understanding that error message after "formal parameters". I just want to know how I can fix the issue, and that message isn't helping.

**Comment by bwilkerson** [comment]
I agree that the message (now messages) are not helpful for understanding the problem or the way to fix it. My response was focused on the steps necessary to fix the messaging.

**Comment by bernaferrari** [comment]
Oh, agreed. You may also try a non-const constructor like `SecondaryConstructor(this.x, this.y);` as the exception would be the same. I think with const it just adds another one on top.

**Comment by jensjoha** [comment]
> If nothing else, this is very poor recovery on the parser's part.

I disagree: the parser itself only gives one error for the file. When adding a third not-quite-constructor without const that doesn't give more errors:
```
class Pixel {
  final int x;
  final int y;

  const Pixel({required this.x, required this.y});

  const SecondaryConstructor(this.x, this.y);
  ThirdConstructor(this.x, this.y);
}
```

still just gives
```
Getters, setters and methods can't be declared to be 'const'.
  const SecondaryConstructor(this.x, this.y); // <-- ERROR, DEFAULT IS ALREADY DEFINED
  ^^^^^
```

Yes, it does say it's a class method (`endClassMethod`) and not a constructor (`endClassConstructor`) (I'd argue that it's not wrong).

We _could_ probably recover more aggressively in the parser, but would then just fall into the opposite trap, saying something like "why didn't you initialize all you're final fields in this constructor?" for a method that says `this.something` in it's parameter list...

Might I suggest adding another tip in for instance the "Field formal parameters can only be used in a constructor" saying something like "if you intend this to be a constructor you should prefix it with the class name".

**Comment by bwilkerson** [comment]
Sorry, I was too hasty in assuming that these diagnostics were primarily coming from the parser.

I still believe that we can improve the UX by reporting fewer diagnostics.

I think there's an interesting question: When the syntactic structure of the code allows us to look at a function as either a constructor with the wrong name or as a method with invalid parameters, which should we choose? I think the right answer is whichever of those most frequently matches the user's intent. I don't have any data, but I suspect that it's more common for users to mistype the name of a constructor than it is for them to try to add field formal parameters to a method (or trying to prefix a method with `const`, for that matter).

Traditionally I believe that we've only used the name of the function (if it matches the name of the class then it's a constructor, if not then it isn't), but I have to ask whether we shouldn't use more than one signal to make that determination.
